### ISSUE TITLE
Convert install script to using new PnP.PowerShell module

### ISSUE THREAD

**Issue by okms** [Puzzlepart/prosjektportalen365#417]
**Description**
PP365 is currently using (and bundling) the unmaintained old version of PnP PowerShell. Ath the moment this is not an issue, but converting to the new [PnP.PowerShell](https://github.com/pnp/powershell) module will ensure that we are ready to support new features as they are available in the graph, and it may even allow us to deploy pp365 from a non-windows computer. 

**Additional context**
There are some breaking changes between the classic PnP PowerShell and the new PnP.PowerShell module, mostly having to do with naming of cmdlets. There are however some issues with authentication which may or may not be deemed a showstopper. This will however be an issue for anyone running PnP PowerShell against their tennant moving forward with the new module.

We also need to update all installation documentation if we go through with this change.

Feature soundtrack: 🎵[You're the best](https://open.spotify.com/track/4zXa17K83Pp6N2yXdVc2sv?si=ILHtSWUlSiSUo8pJvXCFAg)🎵

**Comment by damsleth** [comment]
I have two suggestions of which I think both are doable.
1. Make the script support both versions of PnP, e.g by using module-checking in conjunction with `Set-Alias` in the start of the script (non-exhaustive list of cmdlets below)
``` 
if(Get-Module -Name "PnP.PowerShell"){
    Set-Alias -Name Apply-PnPProvisioningTemplate -Value Invoke-PnPSiteTemplate
    Set-Alias -Name Read-PnPProvisioningTemplate -Value Read-PnPSiteTemplate
    Set-Alias -Name Apply-PnPPTenantTemplate -Value Invoke-PnPTenantTemplate
    Set-Alias -Name Get-PnPProvisioningTemplate -Value Get-PnPSiteTemplate
    Set-Alias -Name Get-PnPUnifiedGroup -Value Get-PnPMicrosoft365Group
    Set-Alias -Name Set-PnPUnifiedGroup -Value Set-PnPMicrosoft365Group
    Set-Alias -Name New-PnPProvisioningTemplateFromFolder -Value New-PnPSiteTemplateFromFolder
}
```    
2. Do stricter version-checking before running. E.g. installation fails with `SharePointPnPPowerShellOnline` v 3.28.2012.0 (December 2020). 
```  
if(Get-Module SharePointPnPPowerShellOnline){
  $minorVersion = (Get-Module SharePointPnPPowerShellOnline).Version.Minor
  if($minorVersion -lt 21 -or $minorVersion -gt 25)
  {
    Write-Warning "Du bruker en usupportert versjon av PnP-PowerShell!"
  }
}
```

**Comment by olemp** [comment]
> I have two suggestions of which I think both are doable.
> 
> 1. Make the script support both versions of PnP, e.g by using module-checking in conjunction with `Set-Alias` in the start of the script (non-exhaustive list of cmdlets below)
> 
> ```
> if(Get-Module -Name "PnP.PowerShell"){
>     Set-Alias -Name Apply-PnPProvisioningTemplate -Value Invoke-PnPSiteTemplate
>     Set-Alias -Name Read-PnPProvisioningTemplate -Value Read-PnPSiteTemplate
>     Set-Alias -Name Apply-PnPPTenantTemplate -Value Invoke-PnPTenantTemplate
>     Set-Alias -Name Get-PnPProvisioningTemplate -Value Get-PnPSiteTemplate
>     Set-Alias -Name Get-PnPUnifiedGroup -Value Get-PnPMicrosoft365Group
>     Set-Alias -Name Set-PnPUnifiedGroup -Value Set-PnPMicrosoft365Group
>     Set-Alias -Name New-PnPProvisioningTemplateFromFolder -Value New-PnPSiteTemplateFromFolder
> }
> ```
> 
> 1. Do stricter version-checking before running. E.g. installation fails with `SharePointPnPPowerShellOnline` v 3.28.2012.0 (December 2020).
> 
> ```
> if(Get-Module SharePointPnPPowerShellOnline){
>   $minorVersion = (Get-Module SharePointPnPPowerShellOnline).Version.Minor
>   if($minorVersion -lt 21 -or $minorVersion -gt 25)
>   {
>     Write-Warning "Du bruker en usupportert versjon av PnP-PowerShell!"
>   }
> }
> ```

We are bundling pnp-powershell in the installation, so I think we should support one of them, not both. If we change I think we should go for `PnP.PowerShell` only - we need to figure out how to handle auth though..

**Comment by okms** [comment]
1.3.0 does support -interactive which is similar-ish to -WebLogin. You still need to give consent to PnP PowerShell in AAD (i think). It's something worth testing out on a dev tenant

**Comment by olemp** [comment]
> 1.3.0 does support -interactive which is similar-ish to -WebLogin. You still need to give consent to PnP PowerShell in AAD (i think). It's something worth testing out on a dev tenant

So I guess we would only support `-Interactive` mode then?

How should we attack this @tarjeieo? Side-by-side Install script using `PnP-PowerShell` for a while? Start a branch where we move 100% over to `PnP.PowerShell`? Support both modules in the same Install.ps1?

**Comment by olemp** [comment]
@damsleth @okms @tarjeieo 

We need to keep and change the URL of the connection somehow using `-Interactive`.

Had to log in 4 times 🗡️ 

![image](https://user-images.githubusercontent.com/7606007/109785872-de8a7600-7c0c-11eb-9c9b-a25217c3c70f.png)

**Comment by okms** [comment]
[...]
> 1. Do stricter version-checking before running. E.g. installation fails with `SharePointPnPPowerShellOnline` v 3.28.2012.0 (December 2020).
> 
> ```
> if(Get-Module SharePointPnPPowerShellOnline){
>   $minorVersion = (Get-Module SharePointPnPPowerShellOnline).Version.Minor
>   if($minorVersion -lt 21 -or $minorVersion -gt 25)
>   {
>     Write-Warning "Du bruker en usupportert versjon av PnP-PowerShell!"
>   }
> }
> ```

We can use require statements to block the script from running if you do not have the correct module or version loaded

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_requires?view=powershell-5.1

for example you can write 

```
#Requires -Version 5.1
#Requires -Modules @{ ModuleName="SharePointPnPPowerShellOnline"; ModuleVersion ="3.21.2005.2" } 
```

![image](https://user-images.githubusercontent.com/170147/110552558-dd3ed900-8137-11eb-8e2b-9101142b2ab2.png)

This will require SharePointPnPPowerShellOnline, minimum version 3.21 to be loaded and at least powershell 5.1. 

This does not check if PnP.PowerShell is loaded, but it will not run unless SharePointPnPPowersShell is loaded and has the correct verision. There are of course a bunch of considerations here
- Dynamically loading a bundeled module
- PnP.PowerShell already being loaded in memory, creating conflicts with SharePointPnPPowerShellOnline 
- Older versions of SharePointPnPPowerShellOnline being loaded in memory, creating conflicts when overloaded with a newer version (might be caught by a Require statement).

Nevertheless, using require statements in any child script should be enforced

**Comment by olemp** [comment]
> [...]
> 
> > 1. Do stricter version-checking before running. E.g. installation fails with `SharePointPnPPowerShellOnline` v 3.28.2012.0 (December 2020).
> > 
> > ```
> > if(Get-Module SharePointPnPPowerShellOnline){
> >   $minorVersion = (Get-Module SharePointPnPPowerShellOnline).Version.Minor
> >   if($minorVersion -lt 21 -or $minorVersion -gt 25)
> >   {
> >     Write-Warning "Du bruker en usupportert versjon av PnP-PowerShell!"
> >   }
> > }
> > ```
> 
> We can use require statements to block the script from running if you do not have the correct module or version loaded
> 
> https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_requires?view=powershell-5.1
> 
> for example you can write
> 
> ```
> #Requires -Version 5.1
> #Requires -Modules @{ ModuleName="SharePointPnPPowerShellOnline"; ModuleVersion ="3.21.2005.2" } 
> ```
> 
> ![image](https://user-images.githubusercontent.com/170147/110552558-dd3ed900-8137-11eb-8e2b-9101142b2ab2.png)
> 
> This will require SharePointPnPPowerShellOnline, minimum version 3.21 to be loaded and at least powershell 5.1.
> 
> This does not check if PnP.PowerShell is loaded, but it will not run unless SharePointPnPPowersShell is loaded and has the correct verision. There are of course a bunch of considerations here
> 
> * Dynamically loading a bundeled module
> * PnP.PowerShell already being loaded in memory, creating conflicts with SharePointPnPPowerShellOnline
> * Older versions of SharePointPnPPowerShellOnline being loaded in memory, creating conflicts when overloaded with a newer version (might be caught by a Require statement).
> 
> Nevertheless, using require statements in any child script should be enforced

Sure about that syntax? Is the version needed?

![image](https://user-images.githubusercontent.com/7606007/110639805-6c87d300-81b0-11eb-93c9-266c9e115378.png)

**Comment by okms** [comment]
I reccomend postponing this a bit as there is a PnP.PowerShell v2 module in development which will be a breaking change from the current PnP.PowerShell module. Keywords are simpler cmdlets, but more of them

![image](https://user-images.githubusercontent.com/170147/135472514-426281a3-3eb4-4873-9973-5da19c747de3.png)

cc: @tarjeieo @olemp

**Comment by tarjeieo** [comment]
This task is not a priority

**Comment by okms** [comment]
I suggest we close this issue and reopen when the time to upgrade comes

**Comment by olemp** [comment]
> I suggest we close this issue and reopen when the time to upgrade comes

Agreed, also closed PR

**Comment by BjornerJF** [comment]
Nice to see that this issue has been reopened, been waiting for this :) 
PnP.PowerShell will give crossplattform and most important the posibilty to use admin account with MFA, by using Interactive. 
Any idea rough estimate for when it will be released?

**Comment by tarjeieo** [comment]
> Nice to see that this issue has been reopened, been waiting for this :) PnP.PowerShell will give crossplattform and most important the posibilty to use admin account with MFA, by using Interactive. Any idea rough estimate for when it will be released?

Hi @BjornerJF - as you probably guessed, we've met some bumps with this one. It's still a task we will need to resolve sooner rather than later, but it's a bit heavier than anticipated.

**Comment by damsleth** [comment]
> @damsleth @okms @tarjeieo
> 
> We need to keep and change the URL of the connection somehow using `-Interactive`.
> 
> Had to log in 4 times 🗡️
> 

This can possibly be solved using `-interactive` once, then saving `Get-PnPAppAuthAccessToken` as a variable and using it in subsequent sharepoint operations.  
Likewise if you need to perform graph operations, you can save `Get-PnPAccessToken` as a variable and use it when doing e.g. `New-PnPMicrosoft365Group`

**Comment by okms** [comment]
`-Interactive` _should_ be cached by the pwsh session. This is why the newer versions of `PnP.PowerShell` has the `-ForceAuthentication` flag

**Comment by olemp** [comment]
> > @damsleth @okms @tarjeieo
> > We need to keep and change the URL of the connection somehow using `-Interactive`.
> > Had to log in 4 times 🗡️
> 
> This can possibly be solved using `-interactive` once, then saving `Get-PnPAppAuthAccessToken` as a variable and using it in subsequent sharepoint operations. Likewise if you need to perform graph operations, you can save `Get-PnPAccessToken` as a variable and use it when doing e.g. `New-PnPMicrosoft365Group`
> 
> example
> 
> ```
> Connect-PnPOnline -Url "https://mytenant.sharepoint.com/sites/SITE_A" -interactive
> $SPToken = Get-PnPAppAuthAccessToken
> $GraphToken = Get-PnPAccessToken
> Invoke-PnPSiteTemplate -Path ./templateSiteA.pnp
> Disconnect-PnPOnline
> Connect-PnPOnline -Url "https://mytenant.sharepoint.com/sites/SITE_B" -AccessToken $SPToken
> Invoke-PnPSiteTemplate -Path ./templateSiteB.pnp
> ```

Seems like the auth token from `Get-PnPAppAuthAccessToken` is per site to we need to store it per URL (so a `HashObject` with 3 entries).

![image](https://user-images.githubusercontent.com/7606007/197204993-74cf85b9-e50c-4632-ad94-77ceb6358b75.png)


FYI @okms @damsleth @Remi749

**Comment by olemp** [comment]
Our script in the Install package `UpgradeAllSitesToLatest.ps1` loops through all projects in the portfolio (hub site), connects to them, and executes different commands. With `PnP.PowerShell` we need to use `-Interactive´ instead of `-UseWebLogin` and it seems like the auth token can only be cached per URL - ergo the user will be prompted to log in for each and every site..

FYI @okms @Remi749

**Comment by okms** [comment]
For future reference PR #42 hit a brick wall due to fields not being provisioned.

**Comment by olemp** [comment]
> For future reference PR #421 hit a brick wall due to fields not being provisioned.

`UpdateChilden="TRUE"` might help. https://github.com/Puzzlepart/prosjektportalen365/commit/fc7d982b16ad678f6a84253303b1fb8ab006f05c

**Comment by olemp** [comment]
> > For future reference PR #421 hit a brick wall due to fields not being provisioned.
> 
> `UpdateChilden="TRUE"` might help. [fc7d982](https://github.com/Puzzlepart/prosjektportalen365/commit/fc7d982b16ad678f6a84253303b1fb8ab006f05c)

Seems like it's working now.

**Comment by olemp** [comment]
Finally fixed in #830 🚀
